report/3.Theory/SymbolTable.tex

```diff
@@ -0,0 +1,60 @@
+%THIS SECTION IS SUPPOSED TO BE A SUBSECTION TO THE PARSER SECTION.
+%Should present the interface, which all symbol table implementations must implement.
+%Should clarify all considerations through the design process.
+\subsection{Symbol Table}
+This section concerns with the implementation of the symbol table. 
+First, the interface implemented by the the symbol table will be presented.
+Second, all considerations through the design process will be discussed.
+This will serve as the reasoning behind the final implementation details.
+
+\textbf{Symbol Table Interface}\\
+The interface consists of five functions as seen in snippet \ref{lis:STInterface}.
+
+\begin{lstlisting}[language=java,label=lis:STInterface,caption=The interface which all symbol table implementations must implement.]
+interface ISymbolTable{
+fun openScope()
+fun closeScope()
+fun enterSymbol(name: String, type: TreeNode.Type): Boolean
+fun retrieveSymbol(name: String): Symbol
+fun declaredLocally(name: String): Boolean
+}
+\end{lstlisting}
+
+The functions on line 2 and 3 concerns with the handling of scopes.
+Scopes are handled by a stack data structure.
+Intuitively, each element on the stack consists of a local symbol table concerning the current scope.
+
+The $openScope()$ function opens a new scope in the symbol table.
+This is done by pushing a new sub-symbol table onto the stack.
+
+The $closeScope()$ function closes the most recently opened sub-symbol table.
+This is done by popping the table from the stack.
```

Comment:
maybe comment on (as discussed with Bent) that we can do this because as long as we know that typed matched, we don't really need to know them in codegen, as C++ should handle this all fine and dandy.